horace_core/sqw/@SQWDnDBase/private/smooth_dnd.m

```diff
@@ -0,0 +1,157 @@
+            error('HORACE:smooth:invalid_arguments', ...
+                'Ambiguous convolution function name')
+        elseif isempty(ishape)
+            error('HORACE:smooth:invalid_arguments', ...
+                ['Function ''',shape,''' is not recognised as an available option'])
+        end
+        shape = shapes(ishape);     % resolve to full name
+    else
+        error('HORACE:smooth:invalid_arguments', ...
+            'Argument ''shape'' must be a character string')
+    end
+else
+    shape = shape_default;
+    ishape= ishape_default;
+end
+
+% Check width parameters against the shape function
+if strcmp(shape,'resolution') && ~(ndim==2 && numel(width)==3)
+    error('HORACE:smooth:invalid_arguments', ...
+        'Smoothing option ''resolution'' only available for 2D data sets with three width parameters')
+elseif ~(isa_size(width,[1,ndim],'double') || isa_size(width,[1,1],'double'))
+    error('HORACE:smooth:invalid_arguments', ...
+        'ERROR: argument ''width'' must be a scalar or vector with length equal to the dimensions of the dataset')
+end
+
+if isa_size(width,[1,1],'double')
+    width = width*ones(1,ndim); % if input is scalar, expand to dimension of dataset
+end
+
+% Catch trivial case of zero dimensional dataset
+if ndim==0
+    dout = din;
+    return
+end
+
+% Get the lengths along each plot axis and convert widths to number of pixels
+if xunit
+    xlen = zeros(1,ndim);
+    for i=1:ndim
+        xlen(i) = (din.p{i}(end)-din.p{i}(1)) / (numel(din.p{i})-1);
+    end
+    if ~strcmp(shape,'resolution')
+        width = width./xlen;
+    else
+        % Need to convert from Ang^-1 and meV to pixels and pixel gradient
+        width(1) = width(1)/xlen(1);
+        width(2) = width(2)*(xlen(1)/xlen(3));
+        width(3) = width(3)/xlen(3);
+    end
+end
+
+
+% Create convolution array
+% --------------------------
+c = shape_handle{ishape}(width);    % use function handles to create matrix - can add further functions above wihout altering remaining code
+
+
+% Smooth data structure
+% ---------------------
+m=warning('off','MATLAB:divideByZero');     % turn off divide by zero messages, saving present state
+cleanup_warning = onCleanup(@() warning(m.state,'MATLAB:divideByZero')); % add cleanup handler in case an error is raised
+
+index = din.npix~=0;                        % elements with non-zero counts
+weight = convn(double(index),c,'same');     % weight function including only points where there is data
+% Not all algorithms correctly ensure s=e=0 for bins where npix=0, so because here it matters
+% so much, explicitly ensure
+signal=din.s; signal(~index)=0;
+signal = convn(signal,c,'same')./weight;     % points with no data (i.e. signal = 0) do not contribute to convolution
+err=din.e; err(~index)=0;
+err = convn(err,c.^2,'same')./(weight.^2);
+
+clear cleanup_warning;     % return to previous divide by zero message state
+
+signal(~index) = 0;     % restore zero signal to those bins with no data
```

Comment:
Presumably signal ends up ~=0 due to the convolution so add `after convolution` so the reader doesn't have to think